File: packages/core/src/scales/scale.ts
import { autoTicks, defaultFormat } from "../axes/ticks.js";
import type { Range, Tick } from "../types.js";

/**
 * A scale maps data-space values to normalized [0,1] and back, and knows how to
 * generate its own "nice" ticks + default label format. `log` tells layers to
 * apply a log10 transform on the GPU.
 */
export interface Scale {
  readonly type: string;
  readonly log: boolean;
  domain: Range;
  norm(value: number): number;
  invert(t: number): number;
  /** Auto ticks appropriate for this scale type. */
  ticks(target?: number): Tick[];
  /** Default label formatter for a tick value. */
  formatTick(value: number): string;
}

export class LinearScale implements Scale {
  readonly type = "linear";
  readonly log = false;
  domain: Range;
  constructor(domain: Range = [0, 1]) {
    this.domain = domain;
  }
  norm(value: number): number {
    const [a, b] = this.domain;
    return b === a ? 0 : (value - a) / (b - a);
  }
  invert(t: number): number {
    const [a, b] = this.domain;
    return a + t * (b - a);
  }
  ticks(target = 6): Tick[] {
    return autoTicks(this.domain[0], this.domain[1], target);
  }
  formatTick(value: number): string {
    return defaultFormat(value);
  }
}

export class LogScale implements Scale {
  readonly type = "log";
  readonly log = true;
  domain: Range;
  constructor(domain: Range = [1, 1000]) {
    this.domain = LogScale.sanitize(domain);
  }
  private static sanitize(d: Range): Range {
    // Log requires strictly positive bounds.
    const lo = d[0] > 0 ? d[0] : 1e-9;
    const hi = d[1] > lo ? d[1] : lo * 10;
    return [lo, hi];
  }
  private get la() {
    return Math.log10(this.domain[0]);
  }
  private get lb() {
    return Math.log10(this.domain[1]);
  }
  norm(value: number): number {
    if (value <= 0) return 0;
    return (Math.log10(value) - this.la) / (this.lb - this.la);
  }
  invert(t: number): number {
    return Math.pow(10, this.la + t * (this.lb - this.la));
  }
  ticks(): Tick[] {
    const [a, b] = this.domain;
    const lo = Math.floor(Math.log10(a));
    const hi = Math.ceil(Math.log10(b));
    const ticks: Tick[] = [];
    for (let e = lo; e <= hi; e++) {
      const base = Math.pow(10, e);
      ticks.push({ value: base });
      for (let m = 2; m <= 9; m++) ticks.push({ value: m * base, minor: true, grid: false });
    }
    return ticks;
  }
  formatTick(value: number): string {
    const e = Math.round(Math.log10(value));
    if (e <= -4 || e >= 5) return `1e${e}`;
    return defaultFormat(value);
  }
}

// Candidate time steps in milliseconds, coarse→fine chosen by target count.
const SECOND = 1000, MINUTE = 60_000, HOUR = 3_600_000, DAY = 86_400_000;
const TIME_STEPS = [
  SECOND, 5 * SECOND, 15 * SECOND, 30 * SECOND,
  MINUTE, 5 * MINUTE, 15 * MINUTE, 30 * MINUTE,
  HOUR, 3 * HOUR, 6 * HOUR, 12 * HOUR,
  DAY, 2 * DAY, 7 * DAY, 30 * DAY, 90 * DAY, 365 * DAY,
];

function pad2(n: number): string {
  return n < 10 ? `0${n}` : `${n}`;
}

/** Linear over epoch-millisecond values, with calendar-aware ticks/labels. */
export class TimeScale implements Scale {
  readonly type = "time";
  readonly log = false;
  domain: Range;
  constructor(domain: Range = [0, DAY]) {
    this.domain = domain;
  }
  norm(value: number): number {
    const [a, b] = this.domain;
    return b === a ? 0 : (value - a) / (b - a);
  }
  invert(t: number): number {
    const [a, b] = this.domain;
    return a + t * (b - a);
  }
  private chooseStep(target: number): number {
    const ideal = (this.domain[1] - this.domain[0]) / target;
    for (const s of TIME_STEPS) if (s >= ideal) return s;
    return TIME_STEPS[TIME_STEPS.length - 1]!;
  }
  ticks(target = 6): Tick[] {
    const [a, b] = this.domain;
    if (!isFinite(a) || !isFinite(b) || a === b) return [];
    const step = this.chooseStep(target);
    const start = Math.ceil(a / step) * step;
    const ticks: Tick[] = [];
    for (let i = 0; i < 1000; i++) {
      const v = start + i * step;
      if (v > b) break;
      ticks.push({ value: v });
    }
    return ticks;
  }
  formatTick(value: number): string {
    const span = this.domain[1] - this.domain[0];
    const d = new Date(value);
    if (span < DAY) return `${pad2(d.getHours())}:${pad2(d.getMinutes())}`;
    if (span < 90 * DAY) return `${d.getMonth() + 1}/${d.getDate()}`;
    return `${d.getFullYear()}`;
  }
}

/**
 * A categorical (factor) axis. Internally it is **linear over a band domain**
 * `[-0.5, n-0.5]`, so factor `i` sits at the band centre and `norm(i)` equals the
 * ordinary linear projection — the GPU transform needs no special handling, and a
 * layer plotting at integer indices lands on band centres automatically.
 * "Categorical-ness" lives only in the ticks/labels (one per factor).
 */
export class CategoricalScale implements Scale {
  readonly type = "categorical";
  readonly log = false;
  factors: string[];
  domain: Range;
  constructor(factors: string[] = []) {
    this.factors = factors;
    const n = factors.length;
    this.domain = n > 0 ? [-0.5, n - 0.5] : [-0.5, 0.5];
  }
  norm(value: number): number {
    const [a, b] = this.domain;
    return b === a ? 0 : (value - a) / (b - a);
  }
  // Continuous linear inverse (satisfies the Scale contract used by pan/box math).
  // The nearest factor index is `Math.round(scale.invert(t))`.
  invert(t: number): number {
    const [a, b] = this.domain;
    return a + t * (b - a);
  }
  ticks(): Tick[] {
    return this.factors.map((f, i) => ({ value: i, label: f, grid: false }));
  }
  formatTick(value: number): string {
    return this.factors[Math.round(value)] ?? "";
  }
}

const MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];

/**
 * An **ordinal time** axis — the standard financial x-axis. Bars are plotted at
 * evenly-spaced integer indices `0..n-1` (like {@link CategoricalScale}, so the
 * domain is the band `[-0.5, n-1+0.5]`), which **collapses market gaps**
 * (weekends, overnight, holidays) instead of leaving blank space the way a real
 * {@link TimeScale} would. Each index carries a timestamp; ticks are placed at
 * natural calendar boundaries (day / month / year, or hour / minute when zoomed
 * in) and subsampled toward the target count, so labels stay readable at any zoom.
 *
 * Plot your candles/bars at `x = 0,1,2,…` and pass the per-bar epoch-ms `times`.
 */
export class OrdinalTimeScale implements Scale {
  readonly type = "ordinal-time";
  readonly log = false;
  times: number[];
  domain: Range;
  constructor(times: ArrayLike<number> = []) {
    this.times = Array.from({ length: times.length }, (_, i) => times[i]!);
    const n = this.times.length;
    this.domain = n > 0 ? [-0.5, n - 0.5] : [-0.5, 0.5];
  }
  norm(value: number): number {
    const [a, b] = this.domain;
    return b === a ? 0 : (value - a) / (b - a);
  }
  invert(t: number): number {
    const [a, b] = this.domain;
    return a + t * (b - a);
  }
  /** Timestamp at a (rounded, clamped) index. */
  private timeAt(i: number): number {
    const n = this.times.length;
    if (n === 0) return 0;
    const k = Math.max(0, Math.min(n - 1, Math.round(i)));
    return this.times[k]!;
  }
  ticks(target = 6): Tick[] {
    const n = this.times.length;
    if (n === 0) return [];
    const i0 = Math.max(0, Math.ceil(this.domain[0] - 1e-9));
    const i1 = Math.min(n - 1, Math.floor(this.domain[1] + 1e-9));
    if (i1 < i0) return [];

    // Calendar levels coarse→fine. Each `key` labels a bar's period; ticks land on
    // the FIRST bar of every period — so gridlines snap to real calendar dates
    // (month starts, Mondays, …) and stay put when you pan (no drift).
    const DAY = 86_400_000;
    const dayKey = (d: Date) => d.getFullYear() * 10000 + d.getMonth() * 100 + d.getDate();
    const weekKey = (d: Date) => {
      const m = new Date(d.getFullYear(), d.getMonth(), d.getDate());
      const dow = (m.getDay() + 6) % 7; // Monday = 0
      return Math.floor((m.getTime() - dow * DAY) / DAY);
    };
    const dayMonth = (d: Date) => `${d.getDate()} ${MONTHS[d.getMonth()]}`;
    const monthLabel = (d: Date) => (d.getMonth() === 0 ? `${d.getFullYear()}` : MONTHS[d.getMonth()]!);
    const levels = [
      { key: (d: Date) => d.getFullYear(), label: (d: Date) => `${d.getFullYear()}` },
      { key: (d: Date) => d.getFullYear() * 4 + Math.floor(d.getMonth() / 3), label: monthLabel },
      { key: (d: Date) => d.getFullYear() * 12 + d.getMonth(), label: monthLabel },
      { key: weekKey, label: dayMonth },
      { key: dayKey, label: dayMonth },
      { key: (d: Date) => dayKey(d) * 4 + Math.floor(d.getHours() / 6), label: hhmm },
      { key: (d: Date) => dayKey(d) * 24 + d.getHours(), label: hhmm },
      { key: (d: Date) => (dayKey(d) * 24 + d.getHours()) * 4 + Math.floor(d.getMinutes() / 15), label: hhmm },
      { key: (d: Date) => (dayKey(d) * 24 + d.getHours()) * 60 + d.getMinutes(), label: hhmm },
    ];
    const periodStarts = (lvl: (typeof levels)[number]): number[] => {
      const out: number[] = [];
      let prev = i0 > 0 ? lvl.key(new Date(this.times[i0 - 1]!)) : NaN;
      for (let i = i0; i <= i1; i++) {
        const k = lvl.key(new Date(this.times[i]!));
        if (k !== prev) { out.push(i); prev = k; }
      }
      return out;
    };
    // Pick the level whose tick count is closest to `target` (penalize over-dense + too-sparse).
    let chosen: { lvl: (typeof levels)[number]; b: number[] } | null = null;
    let bestScore = Infinity;
    for (const lvl of levels) {
      const b = periodStarts(lvl);
      if (b.length === 0) continue;
      const over = b.length > target * 1.5 ? (b.length - target * 1.5) * 3 : 0;
      const sparse = b.length < 2 ? 100 : 0;
      const score = Math.abs(b.length - target) + over + sparse;
      if (score < bestScore) { bestScore = score; chosen = { lvl, b }; }
    }
    if (!chosen) {
      // No calendar change across the view (a few same-minute bars): space evenly.
      const out: Tick[] = [];
      const step = Math.max(1, Math.floor((i1 - i0) / Math.max(1, target - 1)));
      for (let i = i0; i <= i1; i += step) {
        const d = new Date(this.times[i]!);
        out.push({ value: i, label: `${dayMonth(d)} ${hhmm(d)}` });
      }
      return out;
    }
    const c = chosen;
    return c.b.map((i) => ({ value: i, label: c.lvl.label(new Date(this.times[i]!)) }));
  }
  formatTick(value: number): string {
    const d = new Date(this.timeAt(value));
    return `${d.getDate()} ${MONTHS[d.getMonth()]} ${hhmm(d)}`;
  }
}

function hhmm(d: Date): string {
  return `${pad2(d.getHours())}:${pad2(d.getMinutes())}`;
}

export type ScaleType = "linear" | "log" | "time" | "categorical" | "ordinal-time";

export function makeScale(
  type: ScaleType, domain?: Range, factors?: string[], times?: ArrayLike<number>,
): Scale {
  switch (type) {
    case "linear":
      return new LinearScale(domain);
    case "log":
      return new LogScale(domain);
    case "time":
      return new TimeScale(domain);
    case "categorical":
      return new CategoricalScale(factors ?? []);
    case "ordinal-time":
      return new OrdinalTimeScale(times ?? []);
    default:
      throw new Error(`Unknown scale type: ${type as string}`);
  }
}

File: native/src/scale.cpp
#include "scale.hpp"

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <ctime>
#include <limits>

namespace photon {
namespace {

/// Log scales require strictly positive bounds — same clamp as LogScale.sanitize.
void sanitize_log(double& lo, double& hi) {
  if (!(lo > 0.0)) lo = 1e-9;
  if (!(hi > lo)) hi = lo * 10.0;
}

constexpr double kSecond = 1000.0;
constexpr double kMinute = 60000.0;
constexpr double kHour = 3600000.0;
constexpr double kDay = 86400000.0;

/// Candidate time steps in ms, coarse→fine. Mirrors TIME_STEPS in scale.ts.
constexpr double kTimeSteps[] = {
    kSecond, 5 * kSecond, 15 * kSecond, 30 * kSecond,
    kMinute, 5 * kMinute, 15 * kMinute, 30 * kMinute,
    kHour,   3 * kHour,   6 * kHour,    12 * kHour,
    kDay,    2 * kDay,    7 * kDay,     30 * kDay,   90 * kDay, 365 * kDay,
};

const char* const kMonths[] = {"Jan", "Feb", "Mar", "Apr", "May", "Jun",
                               "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"};

/**
 * Local-time calendar fields for an epoch-ms instant.
 *
 * The web core reads these off a JS `Date`, which is local-time throughout
 * (`getHours`, `getMonth`, `getDate`). Matching that matters: a session axis
 * that labelled bars in UTC would put the day boundary in the wrong place for
 * every market outside it.
 */
struct Calendar {
  int year = 1970;
  int month = 0;  // 0-11, like Date#getMonth
  int day = 1;
  int hours = 0;
  int minutes = 0;
  int weekday = 4;  // 0 = Sunday, like Date#getDay
  bool valid = false;
};

Calendar calendar(double epoch_ms) {
  Calendar out;
  if (!std::isfinite(epoch_ms)) return out;
  // Floor rather than truncate so instants before 1970 land on the right second.
  const double seconds = std::floor(epoch_ms / 1000.0);
  if (seconds < static_cast<double>(std::numeric_limits<std::time_t>::min()) ||
      seconds > static_cast<double>(std::numeric_limits<std::time_t>::max())) {
    return out;
  }
  const std::time_t t = static_cast<std::time_t>(seconds);
  std::tm tm{};
#if defined(_WIN32)
  if (localtime_s(&tm, &t) != 0) return out;
#else
  if (localtime_r(&t, &tm) == nullptr) return out;
#endif
  out.year = tm.tm_year + 1900;
  out.month = tm.tm_mon;
  out.day = tm.tm_mday;
  out.hours = tm.tm_hour;
  out.minutes = tm.tm_min;
  out.weekday = tm.tm_wday;
  out.valid = true;
  return out;
}

/// Epoch-ms of local midnight on the given calendar day.
double local_midnight_ms(const Calendar& c) {
  std::tm tm{};
  tm.tm_year = c.year - 1900;
  tm.tm_mon = c.month;
  tm.tm_mday = c.day;
  tm.tm_isdst = -1;  // let the C library decide, as `new Date(y,m,d)` does
  const std::time_t t = std::mktime(&tm);
  return t == static_cast<std::time_t>(-1) ? 0.0 : static_cast<double>(t) * 1000.0;
}

std::string pad2(int n) {
  char buffer[8];
  std::snprintf(buffer, sizeof(buffer), "%02d", n);
  return std::string(buffer);
}

std::string hhmm(const Calendar& c) {
  return pad2(c.hours) + ":" + pad2(c.minutes);
}

std::string day_month(const Calendar& c) {
  return std::to_string(c.day) + " " + kMonths[c.month];
}

std::string month_label(const Calendar& c) {
  return c.month == 0 ? std::to_string(c.year) : std::string(kMonths[c.month]);
}

/// The calendar levels an ordinal-time axis chooses between, coarse→fine.
/// Each `key` labels a bar's period; ticks land on the first bar of each period,
/// so gridlines snap to real calendar dates and stay put while panning.
struct Level {
  long long (*key)(const Calendar&);
  std::string (*label)(const Calendar&);
};

long long key_year(const Calendar& c) { return c.year; }
long long key_quarter(const Calendar& c) { return c.year * 4LL + c.month / 3; }
long long key_month(const Calendar& c) { return c.year * 12LL + c.month; }
long long key_week(const Calendar& c) {
  const int dow = (c.weekday + 6) % 7;  // Monday = 0
  const double midnight = local_midnight_ms(c);
  return static_cast<long long>(std::floor((midnight - dow * kDay) / kDay));
}
long long key_day(const Calendar& c) { return c.year * 10000LL + c.month * 100LL + c.day; }
long long key_six_hours(const Calendar& c) { return key_day(c) * 4 + c.hours / 6; }
long long key_hour(const Calendar& c) { return key_day(c) * 24 + c.hours; }
long long key_quarter_hour(const Calendar& c) {
  return (key_day(c) * 24 + c.hours) * 4 + c.minutes / 15;
}
long long key_minute(const Calendar& c) { return (key_day(c) * 24 + c.hours) * 60 + c.minutes; }

std::string label_year(const Calendar& c) { return std::to_string(c.year); }

const Level kLevels[] = {
    {key_year, label_year},        {key_quarter, month_label},
    {key_month, month_label},      {key_week, day_month},
    {key_day, day_month},          {key_six_hours, hhmm},
    {key_hour, hhmm},              {key_quarter_hour, hhmm},
    {key_minute, hhmm},
};

}  // namespace

double Scale::norm(double value) const {
  if (is_log()) {
    if (value <= 0.0) return 0.0;
    const double la = std::log10(lo);
    const double lb = std::log10(hi);
    if (lb == la) return 0.0;
    return (std::log10(value) - la) / (lb - la);
  }
  if (hi == lo) return 0.0;
  return (value - lo) / (hi - lo);
}

double Scale::invert(double t) const {
  if (is_log()) {
    const double la = std::log10(lo);
    const double lb = std::log10(hi);
    return std::pow(10.0, la + t * (lb - la));
  }
  return lo + t * (hi - lo);
}

bool Scale::set_domain(double new_lo, double new_hi) {
  // A view has to stay representable, and past a certain zoom-out it stops
  // being so: each wheel notch multiplies the span, and after enough of them
  // `hi - lo` overflows to infinity. From that moment norm() divides by
  // infinity, every point in the series projects to the same place, and
  // invert() can no longer produce a finite domain — so zooming back in does
  // not recover. The chart goes blank and stays blank.
  //
  // Refusing the update turns that cliff into a wall: the furthest zoom-out is
  // simply as far as it goes. The endpoints and the span are all checked,
  // because two finite endpoints can still have an infinite difference.
  if (!std::isfinite(new_lo) || !std::isfinite(new_hi) || !std::isfinite(new_hi - new_lo)) {
    return false;
  }
  lo = new_lo;
  hi = new_hi;
  if (is_log()) sanitize_log(lo, hi);
  return true;
}

void Scale::set_band_domain(size_t n) {
  lo = -0.5;
  hi = n > 0 ? static_cast<double>(n) - 0.5 : 0.5;
}

bool Scale::configure(const ph_axis_desc& desc) {
  type = desc.type;

  factors.clear();
  if (type == PH_SCALE_CATEGORICAL && desc.factors && desc.factor_count > 0) {
    factors.reserve(static_cast<size_t>(desc.factor_count));
    for (int32_t i = 0; i < desc.factor_count; ++i) {
      if (!desc.factors[i]) return false;
      factors.emplace_back(desc.factors[i]);
    }
  }

  times.clear();
  if (type == PH_SCALE_ORDINAL_TIME && desc.times && desc.time_count > 0) {
    times.assign(desc.times, desc.times + desc.time_count);
  }

  // The two index-based scales derive their domain from the data they were
  // given; everything else takes the descriptor's domain, or a type-appropriate
  // default when the caller left it empty to mean "autoscale".
  if (type == PH_SCALE_CATEGORICAL) {
    set_band_domain(factors.size());
  } else if (type == PH_SCALE_ORDINAL_TIME) {
    set_band_domain(times.size());
  } else if (desc.domain.lo != desc.domain.hi) {
    set_domain(desc.domain.lo, desc.domain.hi);
  } else if (type == PH_SCALE_LOG) {
    set_domain(1.0, 1000.0);
  } else if (type == PH_SCALE_TIME) {
    set_domain(0.0, 86400000.0);
  } else {
    set_domain(0.0, 1.0);
  }
  return true;
}

// -- ticks ------------------------------------------------------------------

namespace {

/// LogScale.ticks: a decade major plus 2..9 minors, across the visible decades.
std::vector<Tick> log_ticks(double lo, double hi) {
  std::vector<Tick> out;
  const int first = static_cast<int>(std::floor(std::log10(lo)));
  const int last = static_cast<int>(std::ceil(std::log10(hi)));
  // A domain spanning hundreds of decades would emit tens of thousands of ticks.
  if (last - first > 320) return out;
  for (int e = first; e <= last; ++e) {
    const double base = std::pow(10.0, e);
    Tick major;
    major.value = base;
    out.push_back(major);
    for (int m = 2; m <= 9; ++m) {
      Tick minor;
      minor.value = m * base;
      minor.minor = true;
      minor.grid = false;
      out.push_back(minor);
    }
  }
  return out;
}

/// TimeScale.chooseStep: coarsest candidate step that is still fine enough.
double choose_time_step(double span, int target) {
  const double ideal = span / (target > 0 ? target : 1);
  for (const double step : kTimeSteps) {
    if (step >= ideal) return step;
  }
  return kTimeSteps[std::size(kTimeSteps) - 1];
}

std::vector<Tick> time_ticks(double lo, double hi, int target) {
  std::vector<Tick> out;
  if (!std::isfinite(lo) || !std::isfinite(hi) || lo == hi) return out;
  const double step = choose_time_step(hi - lo, target);
  const double start = std::ceil(lo / step) * step;
  for (int i = 0; i < 1000; ++i) {
    const double v = start + i * step;
    if (v > hi) break;
    Tick tick;
    tick.value = v;
    out.push_back(tick);
  }
  return out;
}

}  // namespace

std::vector<Tick> Scale::ticks(int target) const {
  switch (type) {
    case PH_SCALE_LOG:
      return log_ticks(lo, hi);

    case PH_SCALE_TIME:
      return time_ticks(lo, hi, target);

    case PH_SCALE_CATEGORICAL: {
      std::vector<Tick> out;
      out.reserve(factors.size());
      for (size_t i = 0; i < factors.size(); ++i) {
        Tick tick;
        tick.value = static_cast<double>(i);
        tick.label = factors[i];
        tick.grid = false;
        out.push_back(tick);
      }
      return out;
    }

    case PH_SCALE_ORDINAL_TIME:
      return ordinal_time_ticks(target);

    case PH_SCALE_LINEAR:
    default:
      return auto_ticks(lo, hi, target);
  }
}

std::vector<Tick> Scale::ordinal_time_ticks(int target) const {
  std::vector<Tick> out;
  const long long n = static_cast<long long>(times.size());
  if (n == 0) return out;

  long long i0 = static_cast<long long>(std::ceil(lo - 1e-9));
  long long i1 = static_cast<long long>(std::floor(hi + 1e-9));
  i0 = std::max<long long>(0, i0);
  i1 = std::min<long long>(n - 1, i1);
  if (i1 < i0) return out;

  // Precompute the calendar fields once; every level below reads them.
  std::vector<Calendar> cal(static_cast<size_t>(i1 - i0 + 1));
  for (long long i = i0; i <= i1; ++i) {
    cal[static_cast<size_t>(i - i0)] = calendar(times[static_cast<size_t>(i)]);
  }
  const Calendar before = i0 > 0 ? calendar(times[static_cast<size_t>(i0 - 1)]) : Calendar{};

  const auto period_starts = [&](const Level& level) {
    std::vector<long long> starts;
    // Seed with the bar *before* the window so a period already in progress at
    // the left edge does not get a spurious tick.
    bool have_prev = i0 > 0;
    long long prev = have_prev ? level.key(before) : 0;
    for (long long i = i0; i <= i1; ++i) {
      const long long k = level.key(cal[static_cast<size_t>(i - i0)]);
      if (!have_prev || k != prev) {
        starts.push_back(i);
        prev = k;
        have_prev = true;
      }
    }
    return starts;
  };

  // Pick the level whose tick count sits closest to `target`, penalizing
  // over-dense and near-empty results.
  const Level* chosen = nullptr;
  std::vector<long long> chosen_starts;
  double best = std::numeric_limits<double>::infinity();
  for (const Level& level : kLevels) {
    const std::vector<long long> starts = period_starts(level);
    if (starts.empty()) continue;
    const double count = static_cast<double>(starts.size());
    const double over = count > target * 1.5 ? (count - target * 1.5) * 3.0 : 0.0;
    const double sparse = count < 2 ? 100.0 : 0.0;
    const double score = std::fabs(count - target) + over + sparse;
    if (score < best) {
      best = score;
      chosen = &level;
      chosen_starts = starts;
    }
  }

  if (!chosen) {
    // No calendar boundary anywhere in view (a handful of same-minute bars):
    // fall back to even spacing with a full date-and-time label.
    const long long step = std::max<long long>(1, (i1 - i0) / std::max(1, target - 1));
    for (long long i = i0; i <= i1; i += step) {
      const Calendar& c = cal[static_cast<size_t>(i - i0)];
      Tick tick;
      tick.value = static_cast<double>(i);
      tick.label = day_month(c) + " " + hhmm(c);
      out.push_back(tick);
    }
    return out;
  }

  out.reserve(chosen_starts.size());
  for (const long long i : chosen_starts) {
    Tick tick;
    tick.value = static_cast<double>(i);
    tick.label = chosen->label(cal[static_cast<size_t>(i - i0)]);
    out.push_back(tick);
  }
  return out;
}

std::string Scale::format_tick(double value) const {
  switch (type) {
    case PH_SCALE_LOG: {
      if (!(value > 0.0)) return default_format(value);
      const int e = static_cast<int>(std::lround(std::log10(value)));
      if (e <= -4 || e >= 5) return "1e" + std::to_string(e);
      return default_format(value);
    }

    case PH_SCALE_TIME: {
      const double span = hi - lo;
      const Calendar c = calendar(value);
      if (!c.valid) return default_format(value);
      if (span < kDay) return hhmm(c);
      if (span < 90 * kDay) return std::to_string(c.month + 1) + "/" + std::to_string(c.day);
      return std::to_string(c.year);
    }

    case PH_SCALE_CATEGORICAL: {
      const long long i = std::lround(value);
      if (i < 0 || i >= static_cast<long long>(factors.size())) return std::string();
      return factors[static_cast<size_t>(i)];
    }

    case PH_SCALE_ORDINAL_TIME: {
      if (times.empty()) return std::string();
      long long i = std::lround(value);
      i = std::max<long long>(0, std::min<long long>(static_cast<long long>(times.size()) - 1, i));
      const Calendar c = calendar(times[static_cast<size_t>(i)]);
      if (!c.valid) return std::string();
      return day_month(c) + " " + hhmm(c);
    }

    case PH_SCALE_LINEAR:
    default:
      return default_format(value);
  }
}

}  // namespace photon
